[Qiao Yi] iP

meatyturtle opened the pull request:
# Sigma Pro
> "Sigma sigma on the wall, who is the skibidiest of them all?"

SigmaPro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER FAST_ to us

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

What can SigmaPro do?
- [x] Manage tasks
- [x] Manage deadlines
- [ ] Reminders _(COMING SOON!)_

Here's a sneak peek of what SigmaPro does
```java
public void addTask(String task) {
    tasks.add(task);
    saveTasksToFile();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

meatyturtle committed:
Implemented skeletal version

meatyturtle committed:
Implemented Level 1: Echo

meatyturtle committed:
Implemented Level 2: Add items and list items

meatyturtle committed:
Implemented Level 3: Ability to mark and unmark tasks

meatyturtle committed:
Implemented Level 4: Added support for tracking ToDos, Events and Deadlines

meatyturtle committed:
Automated Text UI Testing

meatyturtle committed:
Added support for handling invalid user input

meatyturtle committed:
Implemented Level 8: Ability to delete tasks

meatyturtle committed:
Partially implemented Level 7, listing out tasks from memory but not from dynamic array

meatyturtle committed:
Implemented level 7: ability to save todo list to separate file

meatyturtle committed:
Implemented level 7: ability to save todo list to separate file

meatyturtle committed:
Implemented level 8: chatbot now parses dates and times into friendly format

meatyturtle committed:
used more OOP - extracted methods into appropriate classes

meatyturtle committed:
organised classes into appropriate packages

meatyturtle committed:
merge level 8 branch to master

meatyturtle committed:
use gradle

meatyturtle committed:
resolve merge conflict in branch more oop

meatyturtle committed:
Add JUnit tests for Deadline and TaskList

meatyturtle committed:
Add JavaDoc comments to all non-private classes/methods, and non-trivial private methods

meatyturtle committed:
Comply with Java coding standards

meatyturtle committed:
Implement level 9: Add function to find task by keyword(s)

tckeong commented on src/main/java/Sigma.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Scanner;
+import sigma.*;
```
I think all the packages imported should be list out explicitly, it is more convenient to tracking what packages we use in our program.

tckeong commented on src/main/java/Sigma.java:
```diff
@@ -0,0 +1,87 @@
+
+    /**
+     * Constructor for Sigma
+     * @param filePath The path to the file where past data is stored
```
As the format suggest in the [JavaDocs](https://se-education.org/guides/conventions/java/intermediate.html), I think you should make an empty line between description and parameter section.

tckeong commented on src/main/java/sigma/Parser.java:
```diff
@@ -0,0 +1,188 @@
+import java.util.ArrayList;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import sigma.task.*;
```
I think here can list out all the things imported from sigma.task package

tckeong commented on src/main/java/sigma/Parser.java:
```diff
@@ -0,0 +1,188 @@
+     * @param date The date given by the user as a String
+     * @return The formatted date object
+     */
+    public static String stringToDate(String date) {
```
Follow the suggestion from the [JavaDocs](https://se-education.org/guides/conventions/java/intermediate.html), I think the names of the methods should be start from verb.

tckeong commented on src/main/java/sigma/Parser.java:
```diff
@@ -0,0 +1,188 @@
+            String from = matcher.group(2);
+            String fromDate = stringToDate(from);
+            String to = matcher.group(3);
+            String toDate = stringToDate(to);
```
I think these can combine to single code
```
String toDate = stringToDate(matcher.group(3));
```

tckeong commented on src/main/java/sigma/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+ * @author Qiao Yi
+ */
+public class TaskList {
+    private static List<Task> items;
+
+    /**
+     * Constructor for a TaskList
+     */
+    public TaskList() {
+        items = new ArrayList<>();
+    }
```
Some personal suggestions, not relevant to code styles stuff. I think you can make the constructor become private, so that others cannot use the default constructor, then you can just initialize the items when declare. I think your current implementation will get problem if I call other functions before construct a TaskList, since other functions are static, so this condition may happen.

After changing like this,
```
private static List<Task> items = new ArrayList<>();

private TaskList() {
    // do nothing
    // just prevent other call this constructor 
}
```

tckeong commented on src/test/java/sigma/TaskListTest.java:
```diff
@@ -0,0 +1,74 @@
+package sigma;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import sigma.Parser;
+import sigma.TaskList;
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import static org.junit.jupiter.api.Assertions.*;
+import sigma.task.Task;
```
As the suggestion from the [JavaDocs](https://se-education.org/guides/conventions/java/intermediate.html), I think you can split the imported packages into 3 part, the sigma part, JUnit part and java part

tckeong commented on src/test/java/sigma/TaskListTest.java:
```diff
@@ -0,0 +1,74 @@
+import static org.junit.jupiter.api.Assertions.*;
+import sigma.task.Task;
+public class TaskListTest {
+    TaskList taskList;
+    Parser parser;
```
I think here can explicitly declare the access modifiers of the class variables.

tckeong reviewed (commented):
Overall are good! Just some minor coding styles problems.

DennieDan commented on src/main/java/sigma/Parser.java:
```diff
@@ -0,0 +1,188 @@
+
+    /**
+     * Adds an Event task to the TaskList
+     * @param userInput The Event task that the user wants to add
```
As mentioned by @tckeong, just to remind you to fix the JavaDoc in this file also 😺

DennieDan commented on src/main/java/sigma/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+     * @param userInput The task that the user wants to delete
+     */
+    public void handleDelete(String userInput) {
+        Pattern pattern = Pattern.compile("(delete) (\\d+)");
```
Wow 😮 this is a nice way to parse. I just learned sth, thanks!

DennieDan commented on src/main/java/sigma/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+            if (taskNumber >= 0 && taskNumber < items.size()) {
+                Task task = items.get(taskNumber);
+                items.remove(task);
+                System.out.println("task removed:\n" + task.toString() + "\nNow you have " + items.size() + " tasks in the list");
```
I'm afraid that this line is more than 120 characters.
If you use IntelliJ IDEA, you can try this way to automize wrap text in settings:

Go to **File** -> **Settings** -> **Editor** -> **Code Style** -> **Java**. Under **Hard wrap at**, change to 120.

DennieDan reviewed (commented):
Looks good in general, most of the coding standard minor faults have been pointed out by @tckeong!

Gra7ityIC3 commented on src/main/java/sigma/Storage.java:
```diff
@@ -0,0 +1,71 @@
+package sigma;
+import sigma.task.Task;
```
You should place Java imports first followed by your own imports.

Gra7ityIC3 commented on src/main/java/sigma/task/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+ * @author Qiao Yi
+ */
+public class Deadline extends Task {
+    private String deadline;
```
Perhaps you could make `deadline` final to improve immutability?

Gra7ityIC3 commented on src/main/java/sigma/Storage.java:
```diff
@@ -0,0 +1,71 @@
+        for (String line: Files.readAllLines(Paths.get(path))) {
+            itemsFromFile.append(i).append(". ").append(line).append("\n");
+            i++;
+        }
```
There should be a space before the colon in `for (String line: ...)`.

You could also place `i++` inside `append(i)` like so: `append(i++)`.

Gra7ityIC3 commented on src/main/java/sigma/TaskList.java:
```diff
@@ -0,0 +1,88 @@
+        int i = 1;
+        StringBuilder result = new StringBuilder(); // this is a terrible time complexity
+        for (Task item : items) {
+            result.append("\n").append(i).append(". ").append(item.toString());
```
You could omit the `toString()` call as `StringBuilder::append` implicitly invokes it for objects.

Gra7ityIC3 commented on src/main/java/sigma/task/Event.java:
```diff
@@ -0,0 +1,33 @@
+ * @author Qiao Yi
+ */
+public class Event extends Task {
+    private String startTime;
+    private String endTime;
```
Similarly, you could make `startTime` and `endTime` final to improve immutability.

Gra7ityIC3 commented on src/main/java/sigma/task/ToDo.java:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author Qiao Yi
+ */
+public class ToDo extends Task {
```
Consider renaming `ToDo` to `Todo`, although it's up to your preference!

Gra7ityIC3 commented on src/test/java/sigma/TaskListTest.java:
```diff
@@ -0,0 +1,74 @@
+    TaskList taskList;
+    Parser parser;
+
+    @BeforeEach
+    public void setup() {
+        taskList = new TaskList();
+        parser = new Parser();
+    }
```
Nice use of the `@BeforeEach`  annotation! 👍

Gra7ityIC3 reviewed (commented):
As others have mentioned, you could format your Javadoc comments and organize your imports better. Otherwise, LGTM.

meatyturtle committed:
Add GUI for chatbot

meatyturtle committed:
Use assertions

meatyturtle committed:
Merge pull request #1 from meatyturtle/branch-A-Assertions

Use assertions to verify important assumptions

meatyturtle committed:
Improve code quality

meatyturtle committed:
Merge pull request #2 from meatyturtle/branch-A-CodeQuality

Improve code quality

meatyturtle committed:
Add more flexibility in searching

meatyturtle committed:
Merge branch 'branch-extension'

meatyturtle committed:
Merge https://github.com/meatyturtle/ip

meatyturtle committed:
Fix GUI startup bug

On launch: Sigma does not greet user\n Sigma should greet and welcome user\nAdd additional method to setSigma to display the welcome message\n This way it shows the welcome message on every launch

meatyturtle committed:
Fix GUI startup bug

On launch: Sigma does not greet user
Sigma should greet and welcome user
Add additional method to setSigma to display the welcome message
This way it shows the welcome message on every launch

meatyturtle committed:
Give Sigma more personality

Current replies follow template suggestion >Changed replies to give Sigma more personality

meatyturtle committed:
Merge https://github.com/meatyturtle/ip

meatyturtle committed:
Update Readme

Add user guide to Readme
Current readme is the default project template
Added project-specific features for new users to be familiar with the
chatbot

meatyturtle committed:
Fix major bugs

Fix major bugs

Storage class parses data file incorrectly: extra ']' appears in output,
throws errors when file is empty and incorrectly handles invalid input

Storage needs to parse file contents exactly as the tasks' toString()
method so it can be treated as a task object

Abstracted out method to parse lines in the file, correctly split the
input string and handle each case for different tasks explicitly

Also fixed minor issues relating to string printing and minor error
handling

meatyturtle committed:
Minor formatting and style changes

meatyturtle committed:
Merge pull request #4 from meatyturtle/branch-fix-bugs

Merge major bug fixes to master branch
